File: 62.不同路径.cpp
/*
 * @lc app=leetcode.cn id=62 lang=cpp
 * @lcpr version=30203
 *
 * [62] 不同路径
 */

// @lc code=start
class Solution {
public:
    int uniquePaths(int m, int n) {
        memo.resize(m + 1, vector<int>(n + 1, -1));
        return backtrack(m, n, 0, 0);
    }

private:
    int res = 0;
    vector<vector<int>> memo;
    int backtrack(int m, int n, int r, int c) {
        if (r == m - 1 && c == n - 1) {
            return 1;
        }
        if (r > m - 1 || c > n - 1 || r < 0 || c < 0) {
            return 0;
        }
        if (memo[r][c] != -1) {
            return memo[r][c];
        }
        // 下
        memo[r][c] = backtrack(m, n, r, c + 1) + backtrack(m, n, r + 1, c);
        
        return memo[r][c];
    }

};
// @lc code=end



/*
// @lcpr case=start
// 3\n7\n
// @lcpr case=end

// @lcpr case=start
// 3\n2\n
// @lcpr case=end

// @lcpr case=start
// 7\n3\n
// @lcpr case=end

// @lcpr case=start
// 3\n3\n
// @lcpr case=end

 */

File: 62.不同路径.java
/*
 * @lc app=leetcode.cn id=62 lang=java
 * @lcpr version=30305
 *
 * [62] 不同路径
 */

// @lc code=start

import java.util.Arrays;

class Solution {
    int[][] memo;
    public int uniquePaths(int m, int n) {
        memo = new int[m][n];
        for (int[] row : memo) {
            Arrays.fill(row, 0);
        }
        
        return backtrack(m, n, 0, 0);
    }

    private int backtrack(int m, int n, int i, int j) {
        if (i >= m || j >= n) {
            return 0;
        }
        if (i == m-1 && j == n-1) {
            return 1;
        }
        if (memo[i][j] > 0) {
            return memo[i][j];
        }

        memo[i][j] += backtrack(m, n, i+1, j);
        memo[i][j] += backtrack(m, n, i, j+1);
        return memo[i][j];
    }
}
// @lc code=end



/*
// @lcpr case=start
// 3\n7\n
// @lcpr case=end

// @lcpr case=start
// 3\n2\n
// @lcpr case=end

 */

